File: Exercise_2/Main.py
class Book:
    def __init__(self, title, author, isbn, price, stock):
        self._title = title
        self._author = author
        self._isbn = isbn
        self._price = price
        self._stock = stock

    # Getters
    def get_title(self): return self._title
    def get_author(self): return self._author
    def get_isbn(self): return self._isbn
    def get_stock(self): return self._stock

    # Setter
    def set_stock(self, stock): self._stock = stock

    # Display method
    def display_details(self):
        print(f"Title: {self._title}, Author: {self._author}, ISBN: {self._isbn}, Price: ${self._price}, Stock: {self._stock}", end="")

class EBook(Book):
    def __init__(self, title, author, isbn, price, stock, file_size):
        super().__init__(title, author, isbn, price, stock)
        self._file_size = file_size

    def display_details(self):
        super().display_details()
        print(f", File Size: {self._file_size} MB")

class PrintedBook(Book):
    def __init__(self, title, author, isbn, price, stock, page_count):
        super().__init__(title, author, isbn, price, stock)
        self._page_count = page_count

    def display_details(self):
        super().display_details()
        print(f", Pages: {self._page_count}")

class Library:
    def __init__(self):
        self.books = [] 

    def add_book(self, book):
        self.books.append(book)

    def display_all_books(self):
        print("\n--- Library Catalog ---")
        for book in self.books:
            book.display_details()

    def search_by_title(self, title):
        print(f"\nSearching for title: {title}")
        found = False
        for book in self.books:
            if book.get_title().lower() == title.lower():
                book.display_details()
                found = True
        if not found:
            print("Book not found.")

    
    def search_by_author(self, author):
        print(f"\nSearching for title: {author}")
        found = False
        for book in self.books:
            if book.get_author().lower() == author.lower():
                book.display_details()
                found = True
        if not found:
            print("Book not found.")


    def borrow_book(self, isbn):
        print(f"\nAttempting to borrow ISBN: {isbn}")
        for book in self.books:
            if book.get_isbn() == isbn:
                if book.get_stock() > 0:
                    book.set_stock(book.get_stock() - 1)
                    print(f"Borrow successful! New stock: {book.get_stock()}")
                else:
                    print("Out of stock!")
                return
        print(f"Book with ISBN {isbn} not found.")

# --- MAIN ---
if __name__ == "__main__":
    lib = Library()

    # Add books
    lib.add_book(EBook("Clean Code", "Robert Martin", "1111", 30.0, 5, 2.5))
    lib.add_book(PrintedBook("The Great Gatsby", "F. Scott Fitzgerald", "2222", 15.0, 2, 180))
    lib.add_book(EBook("Python 101", "Jane Doe", "3333", 45.0, 0, 5.0))

    # Display
    lib.display_all_books()

    # Search
    lib.search_by_title("clean code")

    # Borrow
    lib.borrow_book("2222")
    lib.borrow_book("3333")

    # Display again
    lib.display_all_books()

File: Exercise_2/Main.java
package Exercise_2;

import java.util.ArrayList;
import java.util.List;

class Book {
    private String title;
    private String author;
    private String isbn;
    private double price;
    private int stock;

    public Book(String t, String a, String i, double p, int s) {
        this.title = t;
        this.author = a;
        this.isbn = i;
        this.price = p;
        this.stock = s;
    }

    public String getTitle() { return title; }
    public String getAuthor() { return author; }
    public String getISBN() { return isbn; }
    public int getStock() { return stock; }

    public void setStock(int s) { stock = s; }
    public boolean borrow() {
        if (stock > 0) {
            stock--;
            return true;
        }
        return false;
    }

    public void displayDetails() {
        System.out.print("Title: " + title + ", Author: " + author + 
                         ", ISBN: " + isbn + ", Price: $" + price + ", Stock: " + stock);
    }
};

class EBook extends Book {
    private double fileSize;

    public EBook(String t, String a, String i, double p, int s, double fs) {
        super(t, a, i, p, s);
        this.fileSize = fs;
    }

    @Override
    public void displayDetails() {
        super.displayDetails();
        System.out.println(", File Size: " + fileSize + " MB");
    }
};

class PrintedBook extends Book {
    private int pageCount;

    public PrintedBook(String t, String a, String i, double p, int s, int pc) {
        super(t, a, i, pc, s);
        this.pageCount = pc;
    }

    @Override
    public void displayDetails() {
        super.displayDetails();
        System.out.println(", Pages: " + pageCount);
    }
};

class Library {
    private List<Book> books; 

    public Library() {
        books = new ArrayList<>();
    }

    public void addBook(Book book) {
        books.add(book);
    }

    public void displayAllBooks() {
        System.out.println("\n--- Library Catalog ---");
        for (Book b : books) {
            b.displayDetails();
        }
    }

    public void searchByTitle(String title) {
        System.out.println("\nSearching for title: " + title);
        boolean found = false;
        for (Book b : books) {
            if (b.getTitle().equalsIgnoreCase(title)) {
                b.displayDetails();
                found = true;
            }
        }
        if (!found) System.out.println("Book not found.");
    }
    
    public void searchByAuthor(String author) {
        System.out.println("\nSearching for title: " + author);
        boolean found = false;
        for (Book b : books) {
            if (b.getAuthor().equalsIgnoreCase(author)) {
                b.displayDetails();
                found = true;
            }
        }
        if (!found) System.out.println("Book not found.");
    }
    public void borrowBook(String isbn) {
        System.out.println("\nAttempting to borrow ISBN: " + isbn);
        for (Book b : books) {
            if (b.getISBN().equals(isbn)) {
                if (b.getStock() > 0) {
                    b.setStock(b.getStock() - 1);
                    System.out.println("Borrow successful! New stock: " + b.getStock());
                } else {
                    System.out.println("Out of stock!");
                }
                return;
            }
        }
        System.out.println("Book with ISBN " + isbn + " not found.");
    }
}

public class Main {
    public static void main(String[] args) {
        Library lib = new Library();

        // Add books
        lib.addBook(new EBook("Clean Code", "Robert Martin", "1111", 30.0, 5, 2.5));
        lib.addBook(new PrintedBook("The Great Gatsby", "F. Scott Fitzgerald", "2222", 15.0, 2, 180));
        lib.addBook(new EBook("Java Basics", "John Doe", "3333", 45.0, 0, 5.0));

        // Display
        lib.displayAllBooks();

        // Search
        lib.searchByTitle("Clean Code");

        // Borrow
        lib.borrowBook("2222"); // Success
        lib.borrowBook("3333"); // Fail

        // Display again
        lib.displayAllBooks();
    }
}